### PR TITLE
feat: reduce cg memory footprint

### LINKED ISSUE #205: Reduce memory overhead of CUDA graphs
Current CUDA graph wrapper create a static input and static output per call to the model.
In decoder, it may create a bunch of tensors, we may want to limit those creations and try to recycle them.

### LINKED ISSUE #206: no copy of recycled K/V cache in CUDA graphs
Right now, in Cuda Graph, K/V cache is recycled but still copied back on static graph again and again.
We may want to try to avoid that.

### PR DESCRIPTION
**Pull request by pommedeterresautee**
fix #206 
fix #205

### CODE REVIEW

**Comment by pommedeterresautee** [comment]
test pass

```log
================================================================================================== 2849 passed in 8941.45s (2:29:01) ===================================================================================================
```

File: test/test_torchdynamo.py [hunk 1]
```diff
@@ -135,18 +135,46 @@ def test_t5():
...
     processor = WhisperProcessor.from_pretrained("openai/whisper-tiny")
     inputs = torch.load("test/data/whisper_input.pt")
     with torch.inference_mode(), torch.autocast(dtype=torch.float16, cache_enabled=True, device_type="cuda"):
-        predicted_ids = benchmark(model.generate, inputs, min_length=25, max_length=25, num_beams=2, do_sample=False)
+        predicted_ids = benchmark(
+            model.generate, inputs, min_length=25, max_length=25, num_beams=num_beam, do_sample=False
+        )
         transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True, normalize=True)[0]
         assert (
-            transcription == "mister quilter is the apostle of the middle classes and we are glad to welcome his gospel"
```
Comment (gaetansnl) [inline test/test_torchdynamo.py:151]: about this, is it a good idea to use "in" ? it happened that the beggining was correct but not the end

File: src/kernl/optimizer/attention.py [hunk 2]
```diff
@@ -31,8 +42,11 @@ def attention_wrapper(q, k, v, output, sm_scale, is_causal, attention_mask):
 
     # When there is a large difference between those dimensions, our kernel become inefficient
     # (almost no parallelization), so we use pytorch instead
-    if q.size(-2) == 1 and k.size(-2) > 50:
-        attention_reference(q, k, v, output, sm_scale, is_causal=is_causal, attention_mask=attention_mask)
+    if q.size(-2) == 1 and k.size(-2) > 50 and (attention_mask is None) and not is_causal:
```
Comment (gaetansnl) [inline src/kernl/optimizer/attention.py:45]: don't understand "(attention_mask is None) and not is_causal" the condition is only for attention_vec_mat_forward, for attention_reference it's ok

Comment (pommedeterresautee) [reply src/kernl/optimizer/attention.py:45]: you are right, I ll modify it.

Comment (pommedeterresautee) [reply src/kernl/optimizer/attention.py]: done

**Comment by pommedeterresautee** [comment]
test pass after all the commits above

```log

================================================================================================== 2855 passed in 9623.61s (2:40:23) ===================================================================================================

```

File: test/test_torchdynamo.py (hunk 1 above)
Comment (gaetansnl) [reply test/test_torchdynamo.py:151]: I mean using "in" instead of "=="

File: src/kernl/implementations/cuda_graph.py [hunk 3]
```diff
@@ -12,31 +12,77 @@
 #  See the License for the specific language governing permissions and
```
Comment (gaetansnl) [inline src/kernl/implementations/cuda_graph.py:12]: I don't know why this file is in this folder, it should be in optimizer, maybe mistake in previous commits

**Review by gaetansnl — CHANGES_REQUESTED** [review]
minor only

File: src/kernl/optimizer/attention.py [hunk 4]
```diff
@@ -12,32 +12,46 @@
...
     # When tensors are shaped for bmm, first dimension is used for both batch and heads. Our kernel supports tensors
     # with 4 dimensions, so we add another dimension of size 1 for heads.
+
     extend_head = q.dim() == 3
     if extend_head:
-        q = q.view(q.size(0), 1, q.size(-2), q.size(-1))
-        k = k.view(k.size(0), 1, k.size(-2), k.size(-1))
-        v = v.view(v.size(0), 1, v.size(-2), v.size(-1))
-        output = output.view(output.size(0), 1, output.size(-2), output.size(-1))
+        q.unsqueeze_(dim=1)
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/attention.py:38]: Yes on PyTorch the underscore at the end of a method always mean the op is done in place, aka the original object is mutated

File: src/kernl/optimizer/attention.py [hunk 5]
```diff
@@ -12,32 +12,46 @@
...
+        if (attention_mask is None) and (not is_causal):
+            attention_vec_mat_forward(q, k, v, output, sm_scale, is_causal=is_causal, attention_mask=attention_mask)
+        else:
+            attention_reference(q, k, v, output, sm_scale, is_causal=is_causal, attention_mask=attention_mask)
     else:
         attention_forward(q, k, v, output, sm_scale, is_causal=is_causal, attention_mask=attention_mask)
 
     if extend_head:
-        return output.view(q.size(0), q.size(-2), q.size(-1))
+        output.squeeze_(dim=1)
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/attention.py:54]: Same as above

File: src/kernl/implementations/cuda_graph.py (hunk 3 above)
Comment (pommedeterresautee) [reply src/kernl/implementations/cuda_graph.py:12]: done

File: src/kernl/model_optimization.py [hunk 6]
```diff
@@ -29,7 +29,7 @@ def _compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
     return cuda_graphs_wrapper(gm, example_inputs)
 
 
-def optimize_model(original_model: PreTrainedModel) -> None:
+def optimize_model(model: PreTrainedModel) -> None:
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/model_optimization.py:32]: how?

File: src/kernl/implementations/attention_vec_mat.py [hunk 7]
```diff
@@ -181,9 +181,8 @@ def attention_vec_mat_forward(
     assert is_causal is False, "causal mask is not supported"
     assert attention_mask is None, "attention_mask is not supported"
     assert output.shape == q.shape[:3] + k.shape[-1:], f"{output.shape} != {q.shape[:3] + k.shape[-1:]}"
-    if v.stride()[-1] == 1:  # is row major
-        # change layout from row major (default) to col major to make coalesced memory access in Triton kernel
-        v_col_major = v.permute(0, 1, 3, 2).contiguous().permute(0, 1, 3, 2)
-        # mutate v, so its storage is col major
-        v.set_(source=v_col_major)
+    # print("q", q.size(), q.stride(), len(q.untyped_storage()), q.dtype)
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_vec_mat.py]: removed

File: requirements.txt [hunk 8]
```diff
@@ -1,5 +1,5 @@
 triton==2.0.0.dev20221202
-torch== 2.0.0.dev20221214+cu117
+torch==2.0.0.dev20230104+cu117
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply requirements.txt:2]: updated

**Review by gaetansnl — APPROVED** [review]

**Pull request merged**